json_schemas/schema_validator.py

```diff
@@ -0,0 +1,121 @@
+    submitty_validate_schema(testcase, testcase_properties_schema, 'pre_commands', 'pre_commands', descriptive_title='testcase {0} pre-commands'.format(testcase_num))
+    submitty_validate_schema(testcase, testcase_properties_schema, 'single_port_per_container', 'single_port_per_container', descriptive_title='testcase {0} single_port_per_container option'.format(testcase_num))
+    submitty_validate_schema(testcase, testcase_properties_schema, 'use_router', 'use_router', descriptive_title='testcase {0} use_router option'.format(testcase_num))
+    submitty_validate_schema(testcase, testcase_properties_schema, 'title', 'title', descriptive_title='testcase {0} title'.format(testcase_num))
+
+    containers = testcase['containers']
+    container_num = 0
+    #validate each container in the testcase
+    for container in containers:
+      submitty_validate_schema(container, container_schema, None, None, descriptive_title='testcase {0} container {1}'.format(testcase_num, container_num))
+      container_num +=1
+
+    validators = testcase['validation']
+    validator_num = 0
+    #validate each validation object in the testcase
+    for validator in validators:
+      submitty_validate_schema(validator, abstract_validation_object_schema, None, None, descriptive_title='testcase {0} validator {1} (generic)'.format(testcase_num, validator_num))
+      submitty_validate_schema(validator, specific_validation_object_schema, None, None, descriptive_title='testcase {0} validator {1} (specific)'.format(testcase_num, validator_num))
+      validator_num += 1
+
+    testcase_num+=1
+  #Finally, validate the config as a whole.
+  submitty_validate_schema(config_json, schema, None, None, descriptive_title='whole schema')
+
+#We validate a key within a config and a schema. If no key is given, we evaluate the whole config or schema.
+def submitty_validate_schema(config_json, schema, config_json_key, schema_key, descriptive_title='', required=True):
+  
+  my_json_chunk = config_json.get(config_json_key, None) if config_json_key != None else config_json
+  my_schema = schema.get(schema_key, None) if schema_key != None else schema
+
+  #If we are attempting to validate an item not specified in the schema, fail.
+  if my_schema == None:
+    raise Exception('ERROR: There is no specification for {0} in the schema. Please add a specification.'.format(schema_key))
+
+  if my_json_chunk == None:
+    print('WARNING: could not identify {0} ({1})'.format(descriptive_title, config_json_key))
+    return
+ 
+  try:
+    #validate.
+    validate(instance=my_json_chunk, schema=my_schema)
+  except Exception as e:
+    #If the item was required, print an error. Otherwise, the exception will be suppressed later, so don't be loud. 
+    if required:
+      traceback.print_exc()
+    if descriptive_title != '':
+      print('ERROR: {0} was not properly formatted.'.format(descriptive_title))
+    raise Exception(e.message)
```

Comment:
you can change this to be just `raise` and python will re-raise whatever exception put us here.